File: Complementar - Estruturas de dados (Didático)/Algoritmos17/Alg17.java
/*
♦ Algoritmo 17: Algoritmo de Classificação/Ordenação para Estruturas de Dados Lineares Dinâmicas (não Estáticas): Counting Sort (Ordenação por Contagem).
*/

public class Alg17 {
    public static void Alg17(int[] arr) {
        int max = findMaxValue(arr);
        int min = findMinValue(arr);

        int range = max - min + 1;
        int[] countArray = new int[range];
        int[] outputArray = new int[arr.length];

        // Inicializa o array de contagem
        for (int i = 0; i < arr.length; i++) {
            countArray[arr[i] - min]++;
        }

        // Atualiza o array de contagem para ter a soma acumulativa
        for (int i = 1; i < range; i++) {
            countArray[i] += countArray[i - 1];
        }

        // Constrói o array de saída
        for (int i = arr.length - 1; i >= 0; i--) {
            outputArray[countArray[arr[i] - min] - 1] = arr[i];
            countArray[arr[i] - min]--;
        }

        // Copia o array ordenado de volta para o array original
        for (int i = 0; i < arr.length; i++) {
            arr[i] = outputArray[i];
        }
    }

    public static int findMaxValue(int[] arr) {
        int max = arr[0];
        for (int i = 1; i < arr.length; i++) {
            if (arr[i] > max) {
                max = arr[i];
            }
        }
        return max;
    }

    public static int findMinValue(int[] arr) {
        int min = arr[0];
        for (int i = 1; i < arr.length; i++) {
            if (arr[i] < min) {
                min = arr[i];
            }
        }
        return min;
    }

    public static void printArray(int[] arr) {
        for (int num : arr) {
            System.out.print(num + " ");
        }
        System.out.println();
    }

    public static void main(String[] args) {
        int[] arr = {4, 2, 2, 8, 3, 3, 1};
        System.out.println("Array original:");
        printArray(arr);

        Alg17(arr);

        System.out.println("Array ordenado:");
        printArray(arr);
    }
}

File: Complementar - Estruturas de dados (Didático)/Algoritmos17/Alg17.php
<?php
/*
♦ Algoritmo 17: Algoritmo de Classificação/Ordenação para Estruturas de Dados Lineares Dinâmicas (não Estáticas): Counting Sort (Ordenação por Contagem).
*/

function countingSort(&$arr) {
    $max = max($arr);
    $min = min($arr);

    $range = $max - $min + 1;
    $countArray = array_fill(0, $range, 0);
    $outputArray = array_fill(0, count($arr), 0);

    // Inicializa o array de contagem
    foreach ($arr as $value) {
        $countArray[$value - $min]++;
    }

    // Atualiza o array de contagem para ter a soma acumulativa
    for ($i = 1; $i < $range; $i++) {
        $countArray[$i] += $countArray[$i - 1];
    }

    // Constrói o array de saída
    for ($i = count($arr) - 1; $i >= 0; $i--) {
        $outputArray[$countArray[$arr[$i] - $min] - 1] = $arr[$i];
        $countArray[$arr[$i] - $min]--;
    }

    // Copia o array ordenado de volta para o array original
    for ($i = 0; $i < count($arr); $i++) {
        $arr[$i] = $outputArray[$i];
    }
}

function printArray($arr) {
    echo implode(" ", $arr) . PHP_EOL;
}

$arr = [4, 2, 2, 8, 3, 3, 1];
echo "Array original:" . PHP_EOL;
printArray($arr);

countingSort($arr);

echo "Array ordenado:" . PHP_EOL;
printArray($arr);

?>
